packages/core/src/ui/UICanvas.ts

```diff
@@ -0,0 +1,608 @@
+        this._sortDistance = 0;
+        break;
+      case CanvasRenderMode.ScreenSpaceCamera:
+        this._sortDistance = this._distance;
+        break;
+      case CanvasRenderMode.WorldSpace:
+        this._sortDistance = Vector3.distance(cameraPosition, this._transform.worldPosition);
+        break;
+    }
+  }
+
+  /**
+   * @internal
+   */
+  override _onEnableInScene(): void {
+    const entity = this._entity;
+    entity._dispatchModify(EntityModifyFlags.UICanvasEnableInScene);
+    UIUtils.registerUIToGroup(this, UIUtils.getGroupInParents(entity));
+    const rootCanvas = UIUtils.getRootCanvasInParent(entity);
+    UIUtils.registerUIToCanvas(this, rootCanvas);
+    this._setIsRootCanvas(!rootCanvas);
+    UIUtils.registerEntityListener(this);
+    entity._onUIInteractiveChange(this._runtimeRaycastEnable);
+  }
+
+  /**
+   * @internal
+   */
+  override _onDisableInScene(): void {
+    UIUtils.registerUIToGroup(this, null);
+    UIUtils.registerUIToCanvas(this, null);
+    this._setIsRootCanvas(false);
+    UIUtils.unRegisterEntityListener(this);
+    const entity = this._entity;
+    entity._dispatchModify(EntityModifyFlags.UICanvasDisableInScene);
+    entity._onUIInteractiveChange(false);
+  }
+
+  /**
+   * @internal
+   */
+  _raycast(ray: Ray, out: HitResult, distance: number = Number.MAX_SAFE_INTEGER): boolean {
+    const entity = this._entity;
+    const plane = UIRenderer._tempPlane;
+    const transform = entity.transform;
+    const normal = plane.normal.copyFrom(transform.worldForward);
+    plane.distance = -Vector3.dot(normal, transform.worldPosition);
+    const curDistance = ray.intersectPlane(plane);
+    if (curDistance >= 0 && curDistance < distance) {
+      const hitPointWorld = ray.getPoint(curDistance, UIRenderer._tempVec30);
+      const worldMatrixInv = UIRenderer._tempMat;
+      Matrix.invert(this.entity.transform.worldMatrix, worldMatrixInv);
+      const localPosition = UIRenderer._tempVec31;
+      Vector3.transformCoordinate(hitPointWorld, worldMatrixInv, localPosition);
+      if (this._hitTest(localPosition)) {
+        out.distance = curDistance;
+        out.entity = entity;
+        out.component = this;
+        out.normal.copyFrom(normal);
+        out.point.copyFrom(hitPointWorld);
+        return true;
+      }
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * @internal
+   */
+  @ignoreClone
+  _onEntityModify(flag: EntityModifyFlags): void {
+    switch (flag) {
+      case EntityModifyFlags.SiblingIndex:
+        this._rootCanvas && (this._rootCanvas._hierarchyDirty = true);
+        break;
+      case EntityModifyFlags.UICanvasEnableInScene:
+      case EntityModifyFlags.Parent:
+        const rootCanvas = UIUtils.getRootCanvasInParent(this._entity);
+        rootCanvas && (rootCanvas._hierarchyDirty = true);
+        this._setIsRootCanvas(!rootCanvas);
+        UIUtils.registerUIToCanvas(this, rootCanvas);
+        UIUtils.registerEntityListener(this);
```

Comment:
These lines of logic are duplicated with those in _onEnableInScene. Could we encapsulate an additional method to optimize this？